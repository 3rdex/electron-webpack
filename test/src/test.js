import webpack from "webpack"
import * as path from "path"
import { TmpDir } from "temp-file"
import { copy, mkdir, move, writeFile } from "fs-extra-p"
import { randomBytes } from "crypto"

const MemoryFS = require("memory-fs")

const tmpDir = new TmpDir()

afterEach(() => tmpDir.cleanup())

async function doTest(configurationFile, electronWebpackConfuguration) {
  const projectDir = await getMutableProjectDir()
  const finalConfiguration = Object.assign({projectDir}, electronWebpackConfuguration)
  const configuration = await require(`electron-webpack/${configurationFile}`).default({configuration: finalConfiguration, production: true})
  await testWebpack(configuration, projectDir)
}

test("app", () => doTest("webpack.app.config.js"))

test("main production", () => doTest("webpack.main.config.js"))

test("renderer custom sourceDirectory", async () => {
  const projectDir = await getMutableProjectDir()
  await move(path.join(projectDir, "src/renderer"), path.join(projectDir, "customRenderer"))
  const configuration = await require("electron-webpack/webpack.renderer.config.js").default({production: true, configuration: {
    projectDir,
    renderer: {
      sourceDirectory: "customRenderer"
    },
  }})
  await testWebpack(configuration, projectDir)
})

test("main extra entry point and custom source dir", async () => {
  const projectDir = await getMutableProjectDir()
  await move(path.join(projectDir, "src/main"), path.join(projectDir, "customMain"))
  const configuration = await require("electron-webpack/webpack.main.config.js").default({production: true, configuration: {
    projectDir,
    main: {
      extraEntries: ["@/foo.js"],
      sourceDirectory: "customMain"
    },
  }})
  await testWebpack(configuration, projectDir)
})

async function getMutableProjectDir() {
  const projectDir = await tmpDir.getTempDir()
  await copy(path.join(__dirname, "../fixtures"), projectDir)
  return projectDir
}

test("renderer production", async () => {
  const projectDir = await getMutableProjectDir()

  function createTestAsset(dirName) {
    const dir = path.join(projectDir, dirName)
    return mkdir(dir).then(() => writeFile(path.join(dir, "foo.png"), randomBytes(100 * 1024)))
  }

  // size of file must be greater than url-loader limit
  await Promise.all([
    createTestAsset("a"),
    createTestAsset("b"),
  ])

  const configuration = await require("electron-webpack/webpack.renderer.config.js").default({configuration: {projectDir}, production: true})
  await testWebpack(configuration, projectDir)
})

function addCustomResolver(configuration) {
  expect(configuration.resolveLoader).toBeUndefined()
  configuration.resolveLoader = {}
  configuration.resolveLoader.modules = [path.join(__dirname, "../../node_modules"), path.join(__dirname, "../../packages/electron-webpack/node_modules")]
}

test("title true", () => testTitle(true))
test("title false", () => testTitle(false))
test("title null", () => testTitle(null))

async function testTitle(title) {
  const projectDir = path.join(__dirname, "../fixtures")
  const configuration = await require("electron-webpack/webpack.renderer.config.js").default({production: true, minify: false, configuration: {
    projectDir, title}})
  const fs = await testWebpack(configuration, projectDir, false)
  expect(bufferToString(fs.meta(projectDir + "/dist/renderer/index.html")).toString()).toMatchSnapshot()
}

async function testWebpack(configuration, projectDir, checkCompilation = true) {
  if (Array.isArray(configuration)) {
    configuration.forEach(addCustomResolver)
  }
  else {
    addCustomResolver(configuration)
  }

  const fs = new MemoryFS()
  const stats = await new Promise((resolve, reject) => {
    compile(fs, configuration, resolve, reject)
  })

  debugger
  if (checkCompilation) {
    expect(statToMatchObject(stats, projectDir)).toMatchSnapshot()
    expect(bufferToString(fs.meta(projectDir))).toMatchSnapshot()
  }
  return fs
}

function statToMatchObject(stats, projectDir) {
  if (stats.hasErrors()) {
    throw new Error(stats.toJson().errors)
  }

  // skip first 3 lines - Hash, Version and Time
  return stats.toString()
    .split(/\r?\n/)
    .filter(it => {
      const trimmed = it.trim()
      return !trimmed.startsWith("Time:") && !trimmed.startsWith("Hash:") && !trimmed.startsWith("Version:")
    })
    .join("\n")
    .replace(new RegExp(projectDir, "g"), "<project-dir>")
}

function compile(fs, configuration, resolve, reject) {
  const compiler = webpack(configuration)
  compiler.outputFileSystem = fs
  compiler.run((error, stats) => {
    if (error != null) {
      reject(error)
      return
    }

    resolve(stats)
  })
}

function bufferToString(host) {
  for (const key of Object.getOwnPropertyNames(host)) {
    if (key === "") {
      delete host[key]
    }

    const value = host[key]
    if (value == null) {
      continue
    }

    if (Buffer.isBuffer(value)) {
      host[key] = value.toString()
    }
    else if (typeof value === "object") {
      bufferToString(value)
    }
  }
  return host
}
