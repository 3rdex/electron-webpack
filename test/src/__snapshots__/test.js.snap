// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main extra entry point 1`] = `
"      Asset       Size  Chunks             Chunk Names
    main.js  640 bytes       0  [emitted]  main
     foo.js  627 bytes       1  [emitted]  foo
main.js.map  341 bytes       0  [emitted]  main
 foo.js.map  388 bytes       1  [emitted]  foo
   [0] multi ./test/fixtures/src/main/index.js 28 bytes {0} [built]
   [1] ./test/fixtures/src/main/index.js 13 bytes {0} [built]
   [2] ./test/fixtures/src/main/foo.js 33 bytes {1} [built]"
`;

exports[`main extra entry point 2`] = `
Object {
  "dist": Object {
    "main": Object {
      "foo.js": "require(\\"source-map-support/source-map-support.js\\").install();
module.exports=function(a){function b(d){if(c[d])return c[d].exports;var e=c[d]={i:d,l:!1,exports:{}};return a[d].call(e.exports,e,e.exports,b),e.l=!0,e.exports}var c={};return b.m=a,b.c=c,b.d=function(a,c,d){b.o(a,c)||Object.defineProperty(a,c,{configurable:!1,enumerable:!0,get:d})},b.n=function(a){var c=a&&a.__esModule?function(){return a['default']}:function(){return a};return b.d(c,'a',c),c},b.o=function(a,b){return Object.prototype.hasOwnProperty.call(a,b)},b.p='',b(b.s=2)}({2:function(){'use strict';console.log('hi')}});
//# sourceMappingURL=foo.js.map",
      "foo.js.map": "{\\"version\\":3,\\"sources\\":[\\"webpack:///webpack/bootstrap a56a2634ede7c4ea5dd0\\",\\"webpack:///./src/main/foo.js\\"],\\"names\\":[],\\"mappings\\":\\";0BACA,cAMA,IACA,iBAGA,mBACA,CACA,IACA,KAIA,2CAGA,aAGA,OACA,OAIA,IAzBA,eA4BA,MAGA,uBACA,GACA,kCACA,CACA,gBACA,cAGA,OAGA,iBACA,YACA,qBAA2B,UAA0B,UACrD,YAAiC,CAAe,QAChD,EAEA,aADA,QAIA,mBAAsD,wCAA+D,SAGrH,MAGA,wCC7DA,QAAQ,IAAI\\",\\"file\\":\\"foo.js\\",\\"sourceRoot\\":\\"\\"}",
      "main.js": "require(\\"source-map-support/source-map-support.js\\").install();
module.exports=function(a){function b(d){if(c[d])return c[d].exports;var e=c[d]={i:d,l:!1,exports:{}};return a[d].call(e.exports,e,e.exports,b),e.l=!0,e.exports}var c={};return b.m=a,b.c=c,b.d=function(a,c,d){b.o(a,c)||Object.defineProperty(a,c,{configurable:!1,enumerable:!0,get:d})},b.n=function(a){var c=a&&a.__esModule?function(){return a['default']}:function(){return a};return b.d(c,'a',c),c},b.o=function(a,b){return Object.prototype.hasOwnProperty.call(a,b)},b.p='',b(b.s=0)}([function(a,b,c){a.exports=c(1)},function(){'use strict'}]);
//# sourceMappingURL=main.js.map",
      "main.js.map": "{\\"version\\":3,\\"sources\\":[\\"webpack:///webpack/bootstrap a56a2634ede7c4ea5dd0\\"],\\"names\\":[],\\"mappings\\":\\";0BACA,cAMA,IACA,iBAGA,mBACA,CACA,IACA,KAIA,2CAGA,aAGA,OACA,OAIA,IAzBA,eA4BA,MAGA,uBACA,GACA,kCACA,CACA,gBACA,cAGA,OAGA,iBACA,YACA,qBAA2B,UAA0B,UACrD,YAAiC,CAAe,QAChD,EAEA,aADA,QAIA,mBAAsD,wCAA+D,SAGrH,MAGA\\",\\"file\\":\\"main.js\\",\\"sourceRoot\\":\\"\\"}",
    },
  },
}
`;

exports[`main production 1`] = `
"      Asset       Size  Chunks             Chunk Names
    main.js  640 bytes       0  [emitted]  main
main.js.map  341 bytes       0  [emitted]  main
   [0] multi ./test/fixtures/src/main/index.js 28 bytes {0} [built]
   [1] ./test/fixtures/src/main/index.js 13 bytes {0} [built]"
`;

exports[`main production 2`] = `
Object {
  "dist": Object {
    "main": Object {
      "main.js": "require(\\"source-map-support/source-map-support.js\\").install();
module.exports=function(a){function b(d){if(c[d])return c[d].exports;var e=c[d]={i:d,l:!1,exports:{}};return a[d].call(e.exports,e,e.exports,b),e.l=!0,e.exports}var c={};return b.m=a,b.c=c,b.d=function(a,c,d){b.o(a,c)||Object.defineProperty(a,c,{configurable:!1,enumerable:!0,get:d})},b.n=function(a){var c=a&&a.__esModule?function(){return a['default']}:function(){return a};return b.d(c,'a',c),c},b.o=function(a,b){return Object.prototype.hasOwnProperty.call(a,b)},b.p='',b(b.s=0)}([function(a,b,c){a.exports=c(1)},function(){'use strict'}]);
//# sourceMappingURL=main.js.map",
      "main.js.map": "{\\"version\\":3,\\"sources\\":[\\"webpack:///webpack/bootstrap 87944b77f4706cfecf0b\\"],\\"names\\":[],\\"mappings\\":\\";0BACA,cAMA,IACA,iBAGA,mBACA,CACA,IACA,KAIA,2CAGA,aAGA,OACA,OAIA,IAzBA,eA4BA,MAGA,uBACA,GACA,kCACA,CACA,gBACA,cAGA,OAGA,iBACA,YACA,qBAA2B,UAA0B,UACrD,YAAiC,CAAe,QAChD,EAEA,aADA,QAIA,mBAAsD,wCAA+D,SAGrH,MAGA\\",\\"file\\":\\"main.js\\",\\"sourceRoot\\":\\"\\"}",
    },
  },
}
`;

exports[`renderer production 1`] = `
"          Asset       Size  Chunks             Chunk Names
    renderer.js  668 bytes       0  [emitted]  renderer
     styles.css  120 bytes       0  [emitted]  renderer
renderer.js.map  432 bytes       0  [emitted]  renderer
 styles.css.map   87 bytes       0  [emitted]  renderer
     index.html  335 bytes          [emitted]  
   [0] multi <project-dir>/src/renderer/index.js 28 bytes {0} [built]
   [1] <project-dir>/src/renderer/index.js 33 bytes {0} [built]
   [2] <project-dir>/src/renderer/style.css 41 bytes {0} [built]
   [3] ./node_modules/css-loader!<project-dir>/src/renderer/style.css 317 bytes [built]
   [4] ./node_modules/css-loader/lib/css-base.js 2.26 kB [built]
   [5] ./node_modules/style-loader/lib/addStyles.js 8.7 kB [built]
   [6] ./node_modules/style-loader/lib/urls.js 3.01 kB [built]
Child html-webpack-plugin for \\"index.html\\":
         Asset     Size  Chunks  Chunk Names
    index.html  2.78 kB       0  
       [0] ./node_modules/html-loader?minimize=false!/__virtual__/renderer-index.html 278 bytes {0} [built]
Child extract-text-webpack-plugin node_modules/extract-text-webpack-plugin/dist node_modules/css-loader/index.js!../../../..<project-dir>/src/renderer/style.css:
       [0] ./node_modules/css-loader!<project-dir>/src/renderer/style.css 317 bytes {0} [built]
       [1] ./node_modules/css-loader/lib/css-base.js 2.26 kB {0} [built]"
`;

exports[`renderer production 2`] = `
Object {
  "dist": Object {
    "renderer": Object {
      "index.html": "
<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    
    <script>
      
      require(\\"source-map-support/source-map-support.js\\").install()
    </script>
    
  
  <link href=\\"styles.css\\" rel=\\"stylesheet\\"></head>
  <body>
    <div id=\\"app\\"></div>
  <script type=\\"text/javascript\\" src=\\"renderer.js\\"></script></body>
</html>",
      "renderer.js": "module.exports=function(a){function b(d){if(c[d])return c[d].exports;var e=c[d]={i:d,l:!1,exports:{}};return a[d].call(e.exports,e,e.exports,b),e.l=!0,e.exports}var c={};return b.m=a,b.c=c,b.d=function(a,c,d){b.o(a,c)||Object.defineProperty(a,c,{configurable:!1,enumerable:!0,get:d})},b.n=function(a){var c=a&&a.__esModule?function(){return a['default']}:function(){return a};return b.d(c,'a',c),c},b.o=function(a,b){return Object.prototype.hasOwnProperty.call(a,b)},b.p='',b(b.s=0)}([function(a,b,c){a.exports=c(1)},function(a,b,c){'use strict';Object.defineProperty(b,'__esModule',{value:!0});var d=c(2),e=c.n(d)},function(){}]);
//# sourceMappingURL=renderer.js.map",
      "renderer.js.map": "{\\"version\\":3,\\"sources\\":[\\"webpack:///webpack/bootstrap 9cdfd00de0dab203af30\\",\\"webpack:///./src/renderer/index.js\\",\\"webpack:///./src/renderer/style.css\\"],\\"names\\":[],\\"mappings\\":\\"0BACA,cAMA,IACA,iBAGA,mBACA,CACA,IACA,KAIA,2CAGA,aAGA,OACA,OAIA,IAzBA,eA4BA,MAGA,uBACA,GACA,kCACA,CACA,gBACA,cAGA,OAGA,iBACA,YACA,qBAA2B,UAA0B,UACrD,YAAiC,CAAe,QAChD,EAEA,aADA,QAIA,mBAAsD,wCAA+D,SAGrH,MAGA,2IC3DQ,gBCFR\\",\\"file\\":\\"renderer.js\\",\\"sourceRoot\\":\\"\\"}",
      "styles.css": ".f1{background:image(\\"static/a/foo.png\\")}.f2{background:image(\\"static/b/foo.png\\")}
/*# sourceMappingURL=styles.css.map*/",
      "styles.css.map": "{\\"version\\":3,\\"sources\\":[],\\"names\\":[],\\"mappings\\":\\"\\",\\"file\\":\\"styles.css\\",\\"sourceRoot\\":\\"\\"}",
    },
  },
}
`;

exports[`title false 1`] = `" <!DOCTYPE html> <html> <head> <meta charset=utf-8> <script>require(\\"source-map-support/source-map-support.js\\").install()</script> </head> <body> <div id=app></div> <script type=text/javascript src=renderer.js></script></body> </html>"`;

exports[`title null 1`] = `" <!DOCTYPE html> <html> <head> <meta charset=utf-8> <script>require(\\"source-map-support/source-map-support.js\\").install()</script> </head> <body> <div id=app></div> <script type=text/javascript src=renderer.js></script></body> </html>"`;

exports[`title true 1`] = `" <!DOCTYPE html> <html> <head> <meta charset=utf-8> <title>Test</title> <script>require(\\"source-map-support/source-map-support.js\\").install()</script> </head> <body> <div id=app></div> <script type=text/javascript src=renderer.js></script></body> </html>"`;
